### ISSUE TITLE
Unable to configure UrlRewriteFilter with Spring Boot 1.2.5

### ISSUE THREAD

**Issue by lincoln-spiteri** [spring-projects/spring-boot#3627]
I have a problem configuring an instance of the UrlRewriteFilter in Spring Boot after upgrading to 1.2.5 from 1.1.

My configuration bean is set up as follows:

```java
    @Override
    @Bean
    public FilterRegistrationBean filterRegistrationBean() {

        log.info("Adding URL rewriter Bean");

        FilterRegistrationBean filterRegistrationBean = new FilterRegistrationBean();

        filterRegistrationBean.setFilter(new UrlRewriteFilter());
        filterRegistrationBean.setOrder(1);
        filterRegistrationBean.setDispatcherTypes(DispatcherType.REQUEST, DispatcherType.FORWARD);

        //ServletContextResource res = new ServletContextResource(servletContext, "urlrewrite.xml");
        String conf = "/urlrewrite.xml";
        
        filterRegistrationBean.addInitParameter("logLevel", "INFO");
        filterRegistrationBean.addInitParameter("confPath", conf);

        InputStream inputStream = servletContext.getResourceAsStream(conf);

        log.debug("*********** Input stream = {}", inputStream);

        inputStream = ClassLoader.getSystemResourceAsStream(conf);

        log.debug("*********** Input stream = {}", inputStream);
        
        
        return filterRegistrationBean;
    }
```

Internally the filter tries to get an input stream using `ServletContext.getResourceAsStream` and `ClassLoader.getSystemResourceAsStream` (I have replicated this code in my config bean). Both calls return null and the configuration of the filter fails.

I have tried placing the configuration file in the root of the application (under `/src/main/resources`) and a number of sub-directories beneath that (`WEB-INF`, `META-INF` and `static`) however the result is always the same. Since this did work in Spring Boot 1.1 I suspect that there is an issue with Spring Boot and not the `UrlRewriteFilter`.

Any help would be appreciated.

Lincoln

**Comment by dsyer** [comment]
Sagan uses Spring Boot 1.2.5. Here's how the re-write filter is configured: https://github.com/spring-io/sagan/blob/master/sagan-site/src/main/java/sagan/SiteConfig.java#L88.

**Comment by bclozel** [comment]
It works in Sagan, but I had the very same issue this week, with another app.
Keeping the same configuration and switching to `spring-boot-starter-jetty` instead of tomcat worked. I haven't had the time to dig deeper yet.

@lincoln-spiteri could you try with the jetty starter to see if there's a consistent issue?

**Comment by lincoln-spiteri** [comment]
@bclozel switching to jetty clears the issue.
@dsyer I have tried a configuration is very similar to yours but it does not work. Where do you keep the file urlrewrite.xml?

**Comment by bclozel** [comment]
It's kept [here, in `src/main/resources`](https://github.com/spring-io/sagan/blob/master/sagan-site/src/main/resources/urlrewrite.xml).

**Comment by wilkinsona** [comment]
With the file in `src/main/resources`, this, as far as I can tell, only works with 1.1.x if you use `urlrewrite.xml` for the path (note the lack of leading slash). The resource is not found via the servlet context so it falls back to getting a system resource which is successful.

The key difference with Spring Boot 1.2.x is that it uses Tomcat 8. Tomcat 8 requires the path for a servlet context resource to being with a `/`. As far as I can tell, this is a servlet spec requirement that Tomcat is now enforcing. This forces a change to the configuration as using `urlrewrite.xml` causes startup to fail as Tomcat throws an exception. When you use `/urlrewrite.xml` then, as before, the resource is not found via the servlet context but, crucially, the addition of the leading slash breaks the fallback as `getSystemResourceAsStream` needs to be called without the leading `/`.

**Comment by beku8** [comment]
So there's no way to configure it with Tomcat 8, as long as using Spring Boot? @wilkinsona

**Comment by wilkinsona** [comment]
You can cause Tomcat to relax its requirement for a `/` prefix by setting the system property `org.apache.catalina.core.ApplicationContext.GET_RESOURCE_REQUIRE_SLASH` to `false`. I believe that will allow it to fallback to getting the configuration as a system resource.

**Comment by beku8** [comment]
Thanks @wilkinsona, I probably would never have found that.
Interestingly & counter intuitively its actually working when the property is ```true```. Though it produces(which was also the case in Jetty) the following log when ```logLevel=debug```:
```
org.tuckey.web.filters.urlrewrite.UrlRewriteFilter DEBUG: java.net.MalformedURLException: Path urlrewrite.xml does not start with a "/" character

java.net.MalformedURLException: Path urlrewrite.xml does not start with a "/" character
	at org.apache.catalina.core.ApplicationContext.getResource(ApplicationContext.java:548)
	at org.apache.catalina.core.ApplicationContextFacade.getResource(ApplicationContextFacade.java:199)
	at org.tuckey.web.filters.urlrewrite.UrlRewriteFilter.loadUrlRewriterLocal(UrlRewriteFilter.java:277)
	at org.tuckey.web.filters.urlrewrite.UrlRewriteFilter.loadUrlRewriter(UrlRewriteFilter.java:262)
	at org.tuckey.web.filters.urlrewrite.UrlRewriteFilter.init(UrlRewriteFilter.java:253)
	at org.apache.catalina.core.ApplicationFilterConfig.initFilter(ApplicationFilterConfig.java:279)
	at org.apache.catalina.core.ApplicationFilterConfig.<init>(ApplicationFilterConfig.java:109)
	at org.apache.catalina.core.StandardContext.filterStart(StandardContext.java:4574)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5193)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1409)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1399)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```
 But that is totally tolerable, especially we can hide with different logLevel.
Thanks again!

p.s Do you think it could be a breaking change in the future?

**Comment by wilkinsona** [comment]
>  Do you think it could be a breaking change in the future?

I think it's unlikely, but that's really a question for the Tomcat team.

**Comment by nickcodefresh** [comment]
I've got the same issue with Spring Boot v1.3.2, but neither using Jetty or Tomcat with ```org.apache.catalina.core.ApplicationContext.GET_RESOURCE_REQUIRE_SLASH``` set to ```false``` works. Is there another fix you know of, or an alternative to urlrewrite?

**Comment by nickcodefresh** [comment]
Turns out this is only an issue when running using mvn spring-boot:run. Running as a jar works fine. No idea why the spring boot plugin has an issue, but I suspect it's classpath related

**Comment by wilkinsona** [comment]
@nickcodefresh Could you please share a small sample the illustrates the different behaviour between `mvn spring-boot:run` and running as a jar?

**Comment by nickcodefresh** [comment]
We ended up using a string constant as the urlrewrite.xml, rather than a file,  which got us round this issue

**Comment by outofrange** [comment]
I had the same issue when starting with spring-boot:run and fixed it with the following commit, by extending Tuckey's `UrlRewriteFilter` and loading the config file manually.
https://github.com/crowdsupport/crowdsupport/commit/abf5ed531f524cbc858f7d93f1aaf3b9e6c4f331

Not the cleanest solution, but it seems to work.

I'll try to provide a small example in the next few weeks to isolate the problem as much as possible. In case you are reading this and I haven't done it yet, please ping me :)

**Comment by philwebb** [comment]
@outofrange An isolated sample would be super useful. Thanks!

**Comment by outofrange** [comment]
@wilkinsona @philwebb [There you go!](https://github.com/outofrange/sscce/tree/spring-boot/3627)

There isn't really much to it, all it took was this class (and an existing config file at `/src/main/resources/urlrewrite.xml`):

```
@SpringBootApplication
public class SpringBootTuckeyApplication {
    public static void main(String[] args) {
        SpringApplication.run(SpringBootTuckeyApplication.class, args);
    }

    @Bean
    public FilterRegistrationBean tuckeyRegistrationBean() {
        final FilterRegistrationBean registrationBean = new FilterRegistrationBean();

        registrationBean.setFilter(new UrlRewriteFilter());
        registrationBean.addInitParameter("confPath", "urlrewrite.xml");

        return registrationBean;
    }
}
```

Information about how I ran the application and the relevant log message can be found in [README.md](https://github.com/outofrange/sscce/blob/spring-boot/3627/README.md).

**Comment by wilkinsona** [comment]
I wonder if we should add some auto-configuration for `UrlRewriteFilter` that takes care of this automatically?

**Comment by spencergibb** [comment]
I've thought about that as well for Spring Cloud. Everyone wants to use zuul as a general rewrite engine. This would be helpful.

**Comment by bclozel** [comment]
So we could try auto-configure a `UrlRewriteFilter` when `UrlRewriteFilter` is available on the classpath
and a `urlrewrite.xml` file is available. In that case, we could add properties to enable/disable that feature and configure the file location.

To really make that a consistent feature for all containers, in my opinion, we'd have to subclass `UrlRewriteFilter` and allow for other configuration locations. As of 4.0.1, urlrewritefilter tries to load the configuration file using the system classloader (see [this commit](https://github.com/paultuckey/urlrewritefilter/commit/53bc28400e66e970d2506023c04fe5f3cd1eb4cf)) - which is probably why this is working in an embedded JAR and not with maven.

From what I can see, there are a lot of issues requesting other ways to load the configuration in that project bit it seems maintainers favor "extension" for this.

Given the added value for such a feature and the lack of activity in that project, I'm not in favor of investing in that area.

**Comment by bclozel** [comment]
It seems that the recommended way to add other means of loading the configuration is to subclass the filter itself.

So I'm declining this issue for all the reasons in my previous comment.

**Comment by btiernay** [comment]
I too ran into this issue when upgrading to the latest Spring Boot. I also noticed the the mime type of favicon.ico files wasn't showing up. I stepped debug through the code and found that `new ClassPathResource("org/springframework/mail/javamail/mime.types")` didn't exist when checked in `PathExtensionContentNegotiationStrategy`. I have a feeling this is related because the classloader shows:

![image](https://cloud.githubusercontent.com/assets/2717578/18074678/55eb8f9c-6e3c-11e6-84b0-3528c751fe68.png)

I will try to create a small reproducible example to see if that's the case.
